src/codelabs/null-safety.md

```diff
@@ -0,0 +1,321 @@
+assigned a value immediately. For cases like that, use the `late` keyword.
+`late` is how you tell Dart that:
+
+-   You aren't going to assign that field a value right away.
+-   But you *are* going to assign it a value later.
+-   And you'll make sure it's assigned a value *before* it's accessed.
+
+If you declare a field `late` and the field is read before it is assigned a
+value, a `LateInitializationError` is thrown to tell you what went wrong.
+
+### Example: The `late` keyword
+
+Try using the `late` keyword to correct the following code. For a little extra
+fun afterward, try commenting out the line that sets `description`!
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/late_keyword.dart" replace="/late\ String\ description/String description/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+class Meal {
+  String description;
+
+  void setDescription(String str) {
+    description = str;
+  }
+}
+
+void main() {
+  final myMeal = Meal();
+  myMeal.setDescription('Feijoada!');
+  print(myMeal.description);
+}
+```
+
+### Example: Late circular references
+
+The `late` keyword is really helpful for tricky patterns like circular
+references. Here are two objects that need to maintain non-nullable references
+to each other. Try using the `late` keyword to fix this code. Note that you
+don't have to remove `final`. `late` fields can also be `final`: you set `late
+final` values once, and after that they're read-only.
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/late_circular_references.dart" replace="/late\ final\ Team/final Team/g; /late\ final\ Coach/final Coach/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+class Team {
+  final Coach coach;
+}
+
+class Coach {
+  final Team team;
+}
+
+void main() {
+  final myTeam = Team();
+  final myCoach = Coach();
+  myTeam.coach = myCoach;
+  myCoach.team = myTeam;
+
+  print('All done!');
+}
+```
+
+### Exercise: Late and lazy
+
+Here's another pattern that `late` can help with: lazy initialization for
+expensive non-nullable fields. Try running this code without changing it. What
+do you think will change if you make `_cache` a `late` field?
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/late_lazy.dart"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+int _computeValue() {
+  print('Computing value...');
+  return 3;
+}
+
+class CachedValueProvider {
+  final _cache = _computeValue();
+  int get value => _cache;
+}
+
+void main() {
+  print('Calling constructor...');
+  var provider = CachedValueProvider();
+  print('Getting value...');
+  print('The value is ${provider.value}!');
+}
+```
+
+Fun fact: After you add `late` to the declaration of `_cache`, you can move the
+`_computeValue` function into the `CachedValueProvider` class and the code will
+still work! Initialization expressions for `late` fields can use instance
+methods in their initializers.
+
+## Type promotion
```

Comment:
I think this should be above the Late section; promotion is a more basic concept that should be understood first.